src/dawn/SIR/SIR.proto

```diff
@@ -0,0 +1,608 @@
+// @endcode
+// @ingroup sir_proto
+message StencilFunCallExpr {
+  string callee = 1;           // Identifier of the stencil function (i.e callee)
+  repeated Expr arguments = 2; // List of arguments
+  SourceLocation loc = 3;      // Source location
+}
+
+// @brief Stencil function argument
+//
+// A stencil function argument can be a simple dimension (e.g `i`), an offset (e.g `i+1`) or
+// reference an argument of a stencil functions if the current stencil function is called within the
+// scope of another stencil function (nested calls).
+//
+// @ingroup sir_proto
+message StencilFunArgExpr {
+  // Dimension of the argument
+  Dimension dimension = 1;
+
+  // Offset to the dimension
+  int32 offset = 2;
+
+  // Index of the argument of the stencil function in the outer scope. If unused, the value *has* to
+  // be set to -1.
+  //
+  // @b Example:
+  //
+  // The following example shows the usage of `argument_index`. Imagine the following scenario:
+  // @code{.cpp}
+  //   foo(storage a, dimension dir) {
+  //      bar(dir+1, a);
+  //   }
+  // @endcode
+  //
+  // in which the members of `StencilFunArgExpr` for the argument `dir+1` of `bar` would be:
+  //
+  // @code{.cpp}
+  //  - dimension: -1        // We do not know the dimension yet as it depends on `dir` of `foo`
+  //  - offset: +1           // Whatever `dir` will be, we offset it by `+1`
+  //  - argument_index: 1    // We reference the second argument `dir` of `foo` thus assign
+  //                         // the index 1 as we use a 0 based indexing.
+  // @endcode
+  int32 argument_index = 3;
+
+  // Source location
+  SourceLocation loc = 4;
+}
+
+// @brief Access to a variable
+//
+// @ingroup sir_proto
+message VarAccessExpr {
+  string name = 1;        // Name of the variable
+  Expr index = 2;         // Is it an array access (i.e var[2])?
+  bool is_external = 3;   // Is this an access to a external variable (e.g a global)?
+  SourceLocation loc = 4; // Source location
+}
+
+// @brief Access to a field
+//
+// @ingroup sir_proto
+message FieldAccessExpr {
+  // Name of the field
+  string name = 1;
+
+  // Static offsets
+  //
+  // This represent the offset which are know before instantiationg the stencil function. Meaning,
+  // if we have directional or offset arguments, we have to perform a lazy evaluation to compute
+  // the real offset once we know the mapping of the directions (and offsets) to the actual
+  // arguments of the stencil function.
+  repeated int32 offset = 2;
+
+  // Mapping of the directional and offset arguments of the stencil function.
+  //
+  // The `argumentMap` stores an index to the argument list of the stencil function with -1
+  // indicating this argument is unused. The `argumentOffset` holds the parsed offsets of the
+  // direction (or offset).
+  //
+  // @b Example:
+  // Consider the following example which implements an average stencil function :
+  //
+  // @code{.cpp}
+  // foo(storage in, direction dir) {
```

Comment:
call the function avg if we refer to it as avg later on.

And if we do so, should we have the ()/2 there for consistency?

Reply:
Done